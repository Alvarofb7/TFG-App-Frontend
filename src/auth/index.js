export * from "./components/PasswordValidatorBox";
export * from "./pages/LoginPage";
export * from "./layout/HomePage";
