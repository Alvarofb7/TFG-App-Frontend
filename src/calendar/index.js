
export * from "./components/CalendarEvent";
export * from "./components/CalendarModal";
export * from "./components/FabAddNew";
export * from "./components/FabDelete";
export * from "../ui/components/Navbar";


export * from "./pages/CalendarPage";