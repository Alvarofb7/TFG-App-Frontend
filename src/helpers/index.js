
export * from "./calendarLocalizer";
export * from "./convertEventsToDateEvents";
export * from "./convertNotesToDateNotes";
export * from "./getEnvVariables";
export * from "./getMessages";