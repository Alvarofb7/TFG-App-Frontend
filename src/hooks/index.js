
export * from "./useAuthStore";
export * from "./useCalendarStore";
export * from "./useForm";
export * from "./useNoteStore";
export * from "./useUiStore";