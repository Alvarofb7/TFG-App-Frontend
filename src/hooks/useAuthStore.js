import { useDispatch, useSelector } from "react-redux";
import { clearErrorMessage, onChecking, onLogin, onLogout, onLogoutCalendar } from "../store";
import { api } from "../api";
import Swal from "sweetalert2";

export const useAuthStore = () => {

    const dispatch = useDispatch();
    const { status, user, errorMessage } = useSelector(state => state.auth);

     const saveTokenToLocalStorage = async (token) => {
        localStorage.setItem("token", token);
        localStorage.setItem("token-init-date", new Date().getTime());
    }

    const startLogin = async ({ email, password }) => {
        dispatch(onChecking());
        try {
            const { data } = await api.post("/auth", { email, password });

            await saveTokenToLocalStorage(data.token);

            dispatch(onLogin({ name: data.name, uid: data.uid }));
        } catch (error) {
            dispatch(onLogout(error.response?.data?.msg || "Credenciales incorrectas"));
            Swal.fire("Error en registro", error.response?.data?.msg || "Credenciales incorrectas", "error");
            setTimeout(() => {
                dispatch(clearErrorMessage());
            }, 10);
        }
    }

    const startRegister = async ({ name, email, password }) => {
        dispatch(onChecking());
        try {
            const { data } = await api.post("/auth/new", { name, email, password });

            await saveTokenToLocalStorage(data.token);

            dispatch(onLogin({ name: data.name, uid: data.uid }));
        } catch (error) {
            dispatch(onLogout(error.response?.data?.msg || "Error en la creación de la cuenta"));
            Swal.fire("Error en registro", error.response?.data?.msg || "Error en la creación de la cuenta", "error");
            setTimeout(() => {
                dispatch(clearErrorMessage());
            }, 10);
        }
    }

    const checkAuthToken = async () => {
        const token = localStorage.getItem("token");

        if (!token) return dispatch(onLogout());

        try {
            const { data } = await api.get("/auth/renew");

            await saveTokenToLocalStorage(data.token);

            dispatch(onLogin({ name: data.name, uid: data.uid }));
        } catch (error) {
            localStorage.clear();
            dispatch(onLogout());
        }
    }

    const startLogout = () => {
        localStorage.clear();
        dispatch(onLogoutCalendar());
        dispatch(onLogout());
    }

    return {
        //* Propiedades
        errorMessage,
        status,
        user,

        //* Métodos
        checkAuthToken,
        startLogin,
        startLogout,
        startRegister,
    }
}