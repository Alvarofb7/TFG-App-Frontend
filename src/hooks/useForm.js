import { useEffect, useState } from "react";

export const useForm = (initialForm = {}) => {
	const [formState, setFormState] = useState(initialForm);

	useEffect(() => {
		setFormState(initialForm);
	}, [initialForm]);

	const onInputChange = ({ target }) => {
		const { name, value } = target;
		setFormState({
			...formState,
			[name]: value,
		});
	};

	const [lowerValidated, setLowerValidated] = useState(false);
	const [upperValidated, setUpperValidated] = useState(false);
	const [lengthValidated, setLengthValidated] = useState(false);
	const [numberValidated, setNumberValidated] = useState(false);
	const [specialCharValidated, setSpecialCharValidated] = useState(false);

	const onPasswordValidation = ({ target }) => {
		const { name, value } = target;
		setFormState({
			...formState,
			[name]: value,
		});

		const lower = /[a-z]/;
		const upper = /[A-Z]/;
		const length = /(?=^.{6,20}$)/;
		const number = /\d/;
		const special = /[$!%*?-_&]/;

		setLowerValidated(false);
		setUpperValidated(false);
		setLengthValidated(false);
		setNumberValidated(false);
		setSpecialCharValidated(false);

		if (lower.test(value)) {
			setLowerValidated(true);
		}
		if (upper.test(value)) {
			setUpperValidated(true);
		}
		if (length.test(value)) {
			setLengthValidated(true);
		}
		if (number.test(value)) {
			setNumberValidated(true);
		}
		if (special.test(value)) {
			setSpecialCharValidated(true);
		}
	};

	return {
		//* Variables
		...formState,
		formState,
		lowerValidated,
		upperValidated,
		lengthValidated,
		numberValidated,
		specialCharValidated,

		//* Methos
		onInputChange,
		onPasswordValidation,
	};
};
