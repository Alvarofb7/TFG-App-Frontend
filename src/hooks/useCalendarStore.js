import { useDispatch, useSelector } from "react-redux";
import { onAddNewEvent, onClearActiveEvent, onDeleteEvent, onSetActiveEvent, onUpdateEvent, onLoadEvents } from "../store";
import Swal from "sweetalert2";
import { api } from "../api";
import { convertEventsToDateEvents } from "../helpers";

export const useCalendarStore = () => {

    const dispatch = useDispatch();
    const { events, activeEvent } = useSelector(state => state.calendar);
    const { user } = useSelector(state => state.auth);

    const startLoadingEvents = async () => {
        try {
            const { data } = await api.get("/calendar");

            const events = convertEventsToDateEvents(data.eventos);
            dispatch(onLoadEvents(events));
        } catch (error) {
            console.log(error);
            Swal.fire("Error al cargar los eventos", "No se han podido cargar los elementos", "error");
        }
    }

    const startSavingEvent = async (calendarEvent) => {
        try {
            if (calendarEvent.id) {
                //Actualizando
                await api.put(`/calendar/${calendarEvent.id}`, calendarEvent);

                dispatch(onUpdateEvent({ ...calendarEvent }));
            } else {
                // Creando
                const { data } = await api.post("/calendar", calendarEvent);

                dispatch(onAddNewEvent({ ...calendarEvent, id: data.evento.id, user }));
            }
        } catch (error) {
            console.log(error);
            Swal.fire("Error al guardar", error.response?.data?.msg || "No se ha podido crear/actualizar el evento", "error");
        }
    };

    const startDeletingEvent = async () => {
        try {
            await api.delete(`/calendar/${activeEvent.id}`);

            dispatch(onDeleteEvent());
        } catch (error) {
            console.log(error);
            Swal.fire("Error al eliminar", error.response.data?.msg || "No se ha podido eliminar el evento", "error");
        }
    };

    const setActiveEvent = (calendarEvent) => {
        dispatch(onSetActiveEvent(calendarEvent));
    };

    const quitActiveEvent = () => {
        dispatch(onClearActiveEvent());
    }

    return {
        //* Propiedades
        activeEvent,
        events,
        hasEventSelected: !!activeEvent?.id,

        //* MÃ©todos
        quitActiveEvent,
        startDeletingEvent,
        setActiveEvent,
        startLoadingEvents,
        startSavingEvent,
    }
}


