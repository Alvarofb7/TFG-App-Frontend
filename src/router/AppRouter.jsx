import { Routes, Route, Navigate } from "react-router-dom";

import { LoginPage } from "../auth";
import { CalendarPage } from "../calendar";
import { KanbanPage } from "../kanban";
import { NotesPage } from "../notes";
import { useAuthStore } from "../hooks";

export const AppRouter = () => {
	const { status } = useAuthStore();

	return (
		<>
			<Routes>
				{status !== "authenticated" ? (
					<>
						<Route path="/home" element={<LoginPage />} />

						<Route path="/*" element={<Navigate to="/home" />} />
					</>
				) : (
					<>
						<Route path="/calendar" element={<CalendarPage />} />
						<Route path="/notes" element={<NotesPage />} />
						<Route path="/kanban" element={<KanbanPage />} />

						<Route path="/*" element={<Navigate to="/calendar" />} />
					</>
				)}
			</Routes>
		</>
	);
};
