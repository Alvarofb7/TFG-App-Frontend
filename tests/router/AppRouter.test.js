import { MemoryRouter } from "react-router-dom";
import { AppRouter } from "../../src/router/AppRouter";
import { render, screen } from "@testing-library/react";
import { useAuthStore } from "../../src/hooks";

jest.mock("../../src/hooks/useAuthStore");

jest.mock("../../src/calendar", () => ({
	CalendarPage: () => <h1>CalendarPage</h1>,
}));

describe("Pruebas de AppRouter", () => {
	const mockCheckAuthToken = jest.fn();

	beforeEach(() => jest.clearAllMocks());

	test("Debe de mostrar la pÃ¡gina de home en caso de no estar autenticados", () => {
		useAuthStore.mockReturnValue({
			status: "not-authenticated",
			checkAuthToken: mockCheckAuthToken,
		});

		const { container } = render(
			<MemoryRouter initialEntries={["/home"]}>
				<AppRouter />
			</MemoryRouter>
		);

		expect(screen.getByText("Ingreso")).toBeTruthy();
		expect(container).toMatchSnapshot();
	});

	test("Debe de mostrar el calendario en caso de estar autenticados", () => {
		useAuthStore.mockReturnValue({
			status: "authenticated",
			checkAuthToken: mockCheckAuthToken,
		});

		const { container } = render(
			<MemoryRouter>
				<AppRouter />
			</MemoryRouter>
		);

		expect(screen.getByText("CalendarPage")).toBeTruthy();
		expect(container).toMatchSnapshot();
	});
});
