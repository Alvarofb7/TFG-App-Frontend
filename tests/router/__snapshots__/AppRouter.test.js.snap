// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de AppRouter Debe de mostrar el calendario en caso de estar autenticados 1`] = `
<div>
  <h1>
    CalendarPage
  </h1>
</div>
`;

exports[`Pruebas de AppRouter Debe de mostrar la página de home en caso de no estar autenticados 1`] = `
<div>
  <div
    class="homePage"
    id="home"
  >
    <header>
      <nav
        class="nav"
        id="nav"
      >
        <div
          class="nav-center"
        >
          <div
            class="nav-header"
          >
            <img
              alt="Logo app"
              class="logo"
              src="/img/Logo.png"
            />
            <button
              class="nav-toggle"
            >
              <i
                class="fas fa-bars"
              />
            </button>
          </div>
          <div
            class="links-container"
          >
            <ul
              class="links"
            >
              <li>
                <a
                  class="scroll-link"
                  href="#home"
                >
                  Home
                </a>
              </li>
              <li>
                <a
                  class="scroll-link"
                  href="#features"
                >
                  Funcionalidades
                </a>
              </li>
              <li>
                <a
                  class="scroll-link"
                  href="#about"
                >
                  ¿Quién soy?
                </a>
              </li>
              <li>
                <a
                  class="scroll-link"
                  href="#authentication"
                >
                  Autenticación
                </a>
              </li>
            </ul>
          </div>
        </div>
      </nav>
      <div
        class="banner"
      >
        <div
          class="container-banner"
        >
          <h1>
            TFG App
          </h1>
          <h4>
            La mejor aplicación para organizar tu tiempo y aumentar tu productividad.
          </h4>
        </div>
      </div>
    </header>
    <section
      class="section"
      id="features"
    >
      <div
        class="home-title"
      >
        <h2>
          ¿Qué ofrecemos?
        </h2>
      </div>
      <div
        class="features-slider"
      >
        <div
          class="slider-container"
        >
          <div
            class="slide"
          >
            <img
              alt="Imagen calendario"
              class="slide-img"
              src="/img/Calendario.jpg"
            />
          </div>
          <div
            class="slide"
          >
            <img
              alt="Imagen notas"
              class="slide-img"
              src="/img/Notas.jpg"
            />
          </div>
          <div
            class="slide"
          >
            <img
              alt="Imagen kanban"
              class="slide-img"
              src="/img/Kanban.jpg"
            />
          </div>
        </div>
        <div
          class="btn-container"
        >
          <button
            class="prevBtn"
            type="button"
          >
            <i
              class="fa-solid fa-arrow-left"
            />
          </button>
          <button
            class="nextBtn"
            type="button"
          >
            <i
              class="fa-solid fa-arrow-right"
            />
          </button>
        </div>
      </div>
    </section>
    <section
      class="section-about"
      id="about"
    >
      <div
        class="home-title"
      >
        <h2>
          ¿Quién soy?
        </h2>
      </div>
      <div
        class="about-center section-center"
      >
        <article
          class="about-img"
        >
          <img
            alt="Imagen perfil"
            src="/img/ImagenPerfil.jpg"
          />
        </article>
        <article
          class="about"
        >
          <div
            class="btn-container-about"
          >
            <button
              class="tab-btn active"
              data-id="personal"
            >
              Personal
            </button>
            <button
              class="tab-btn"
              data-id="tfg"
            >
              Trabajo de fin de grado
            </button>
          </div>
          <div
            class="about-content"
          >
            <div
              class="content active"
              id="personal"
            >
              <p>
                Estudiante del último curso de Ingeniería del Software en Sevilla a falta de la realización del trabajo de fin de grado para completarlo.
                <br />
                <br />
                Trabajando desde el 1 de diciembre de 2020 como desarrollador Backend. Realizando tareas de mantenimiento, de desarrollo y análisis tanto funcional como técnico.
                <br />
                <br />
                Competitivo y con muchas ganas de seguir aprendiendo para ser mejor.
              </p>
            </div>
            <div
              class="content"
              id="tfg"
            >
              <p>
                Realizando este trabajo de fin de grado cuyo tema a tratar es el aprendizaje desde cero de la librería React. Para ello se ha necesitado un aprendizaje previo de JavaScript, dado que mi conocimiento era nulo, y luego un aprendizaje de React. Además, he utilizado el stack tecnológico MERN, compuesto de MongoDB, Express, React y Node.js.
                <br />
                <br />
                Tras todo este aprendizaje, he desarrollado una aplicación con un sistema de autenticación robusto y 3 funcionalidades: calendario, notas y Kanban.
              </p>
            </div>
          </div>
        </article>
      </div>
    </section>
    <section
      class="section"
      id="authentication"
    >
      <div
        class="home-title"
      >
        <h2>
          Autenticacion
        </h2>
      </div>
      <div
        class="authenticacion-form"
      >
        <div
          class="container login-container"
        >
          <div
            class="row login-row"
          >
            <div
              class="col-md-6 login-form-1"
            >
              <h3>
                Ingreso
              </h3>
              <form>
                <div
                  class="form-group mb-2"
                >
                  <input
                    class="form-control"
                    name="loginEmail"
                    placeholder="Correo"
                    required=""
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="form-group mb-2 password-input-login"
                >
                  <input
                    class="form-control"
                    name="loginPassword"
                    placeholder="Contraseña"
                    required=""
                    type="password"
                    value=""
                  />
                  <button
                    class="btn btn-outline-primary btn-password-login"
                    type="button"
                  >
                    <i
                      class="fa-regular fa-eye"
                    />
                  </button>
                </div>
                <div
                  class="d-grid gap-2"
                >
                  <button
                    aria-label="btn-login"
                    class="btn btn-light btnSubmit"
                    type="submit"
                  >
                    Login
                  </button>
                </div>
              </form>
            </div>
            <div
              class="col-md-6 login-form-2"
            >
              <h3>
                Registro
              </h3>
              <form>
                <div
                  class="form-group mb-2"
                >
                  <input
                    class="form-control"
                    name="registerName"
                    placeholder="Nombre"
                    required=""
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="form-group mb-2"
                >
                  <input
                    class="form-control"
                    name="registerEmail"
                    placeholder="Correo"
                    required=""
                    type="email"
                    value=""
                  />
                </div>
                <div
                  class="form-group mb-2 password-input-register"
                >
                  <div
                    class="password-register"
                  >
                    <input
                      class="form-control"
                      name="registerPassword"
                      placeholder="Contraseña"
                      required=""
                      type="password"
                      value=""
                    />
                    <button
                      class="btn btn-outline-light btn-password-register"
                      type="button"
                    >
                      <i
                        class="fa-regular fa-eye"
                      />
                    </button>
                  </div>
                  <main
                    aria-label="password-validation-box"
                    class="password-validation-box"
                    style="display: none;"
                  >
                    <div
                      class="not-validated"
                    >
                      <span
                        class="list-icon"
                      >
                        <i
                          class="fa-solid fa-xmark"
                        />
                      </span>
                      Debe de contener entre 6 y 20 caracteres
                    </div>
                    <div
                      class="not-validated"
                    >
                      <span
                        class="list-icon"
                      >
                        <i
                          class="fa-solid fa-xmark"
                        />
                      </span>
                      Al menos una letra minúscula
                    </div>
                    <div
                      class="not-validated"
                    >
                      <span
                        class="list-icon"
                      >
                        <i
                          class="fa-solid fa-xmark"
                        />
                      </span>
                      Al menos una letra mayúscula
                    </div>
                    <div
                      class="not-validated"
                    >
                      <span
                        class="list-icon"
                      >
                        <i
                          class="fa-solid fa-xmark"
                        />
                      </span>
                      Al menos un número
                    </div>
                    <div
                      class="not-validated"
                    >
                      <span
                        class="list-icon"
                      >
                        <i
                          class="fa-solid fa-xmark"
                        />
                      </span>
                      Al menos un caracter especial
                    </div>
                  </main>
                </div>
                <div
                  class="form-group mb-2"
                >
                  <input
                    class="form-control"
                    name="registerPassword2"
                    placeholder="Repita la contraseña"
                    required=""
                    type="password"
                    value=""
                  />
                </div>
                <div
                  class="d-grid gap-2"
                >
                  <button
                    aria-label="btn-create-account"
                    class="btn btn-light btnSubmit"
                    type="submit"
                  >
                    Crear cuenta
                  </button>
                </div>
              </form>
            </div>
          </div>
        </div>
      </div>
    </section>
    <footer
      class="footer"
    >
      <p>
        Derechos reservados TFG App © 2023
      </p>
    </footer>
    <a
      class="scroll-link top-link"
      href="#home"
    >
      <i
        class="fas fa-arrow-up"
      />
    </a>
  </div>
</div>
`;
